File: backend/tavolo.py
"""Il tavolo: un grafo che persona e modello costruiscono insieme.

Due regole tengono in piedi lo strumento.

La prima riguarda il vocabolario. Le connessioni dicono quattro generi di cosa
(un argomento, una causa, un tempo, un'appartenenza), ma il modello nomina un
token solo: la famiglia si deriva qui. Quattro famiglie messe su una lista
piatta farebbero tredici tipi da tenere a mente, e la lezione delle quattro
forme dei nodi vale anche per gli archi. Forza e incertezza restano fuori dal
vocabolario perche' non sono generi di legame: sono aggettivi di qualunque
legame, e messi fra i tipi li avrebbero moltiplicati.

La seconda riguarda chi scrive. Il modello non tocca il tavolo: propone. Una
proposta entra nel grafo con `state="pending"` e non conta per nessuno finche'
la persona non l'accetta. `live()` e' l'unica vista che vale come contenuto —
la resa a parole, l'export e il conteggio passano tutti di li'.
"""
from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, ConfigDict, Field, ValidationError, field_validator

from .diagram_render import DEFAULT_FORM, FORM_FROM_ROLE, NODE_FORMS

FEATURE_KEY = "feature_tavolo"

# Un tavolo non e' un'illustrazione: cresce per un'ora di lavoro, non per un
# messaggio. I tetti sono quelli oltre i quali smette di essere leggibile.
MAX_NODES = 40
MAX_EDGES = 60
MAX_LABEL = 80
MAX_TITLE = 80
# Quante mosse il modello puo' proporre in una volta. Oltre, la persona non
# giudica piu' le proposte una per una: le accetta in blocco o le scarta tutte.
MAX_PROPOSED = 6

FAMILIES = ("argument", "cause", "time", "part")

# Il vocabolario. La chiave e' l'unica cosa che il modello scrive; il valore e'
# il canale visivo, che sceglie il colore del tratto.
REL_FAMILY: dict[str, str] = {
    "supports": "argument",
    "contradicts": "argument",
    "assumes": "argument",
    "needs-evidence": "argument",
    "causes": "cause",
    "hinders": "cause",
    "feeds-back": "cause",
    "then": "time",
    "blocks": "time",
    "if": "time",
    "part-of": "part",
    "example-of": "part",
}

# Come un legame gia' scritto con le convenzioni dei diagrammi entra nel tavolo.
# La mappa di Idea usa `link` come filo dell'albero, non come relazione
# generica: li' un figlio e' un pezzo del ramo che lo tiene.
REL_FROM_EDGE_KIND = {
    "drives": "causes",
    "strengthens": "supports",
    "weakens": "hinders",
    "feedback": "feeds-back",
    "link": "part-of",
    "unclear": "part-of",
}

# Il legame letto a voce. Registro piano come `diagram_render.describe`: serve a
# chi ascolta invece di guardare, non al pannello.
REL_WORDS = {
    "it": {"supports": "sostiene", "contradicts": "contraddice", "assumes": "da' per scontato",
           "needs-evidence": "chiede una prova a", "causes": "porta a", "hinders": "ostacola",
           "feeds-back": "torna su", "then": "viene prima di", "blocks": "blocca",
           "if": "vale se", "part-of": "fa parte di", "example-of": "e' un caso di"},
    "en": {"supports": "supports", "contradicts": "contradicts", "assumes": "takes for granted",
           "needs-evidence": "needs evidence for", "causes": "leads to", "hinders": "hinders",
           "feeds-back": "feeds back into", "then": "comes before", "blocks": "blocks",
           "if": "holds if", "part-of": "is part of", "example-of": "is a case of"},
    "es": {"supports": "sostiene", "contradicts": "contradice", "assumes": "da por supuesto",
           "needs-evidence": "pide una prueba a", "causes": "lleva a", "hinders": "dificulta",
           "feeds-back": "vuelve a", "then": "viene antes de", "blocks": "bloquea",
           "if": "vale si", "part-of": "forma parte de", "example-of": "es un caso de"},
    "fr": {"supports": "soutient", "contradicts": "contredit", "assumes": "tient pour acquis",
           "needs-evidence": "demande une preuve a", "causes": "mene a", "hinders": "entrave",
           "feeds-back": "revient sur", "then": "vient avant", "blocks": "bloque",
           "if": "vaut si", "part-of": "fait partie de", "example-of": "est un cas de"},
    "de": {"supports": "stutzt", "contradicts": "widerspricht", "assumes": "setzt voraus",
           "needs-evidence": "verlangt einen Beleg fur", "causes": "fuhrt zu", "hinders": "behindert",
           "feeds-back": "wirkt zuruck auf", "then": "kommt vor", "blocks": "blockiert",
           "if": "gilt wenn", "part-of": "ist Teil von", "example-of": "ist ein Fall von"},
    "sv": {"supports": "stodjer", "contradicts": "motsager", "assumes": "tar for givet",
           "needs-evidence": "kraver belagg for", "causes": "leder till", "hinders": "hindrar",
           "feeds-back": "aterverkar pa", "then": "kommer fore", "blocks": "blockerar",
           "if": "galler om", "part-of": "ar en del av", "example-of": "ar ett fall av"},
}

# Il dubbio detto a voce. Un legame ipotetico si vede tratteggiato: chi ascolta
# non ha il tratteggio, e senza questa parola sente una certezza.
DOUBT_WORD = {
    "it": "ipotesi", "en": "hypothesis", "es": "hipotesis",
    "fr": "hypothese", "de": "Annahme", "sv": "antagande",
}

# Quanto pesa il legame, per chi ascolta. Il peso medio non si dice: dirlo a
# ogni arco riempirebbe la lettura di parole che non distinguono niente.
STRENGTH_WORD = {
    "it": {1: "a volte", 3: "sempre"},
    "en": {1: "sometimes", 3: "always"},
    "es": {1: "a veces", 3: "siempre"},
    "fr": {1: "parfois", 3: "toujours"},
    "de": {1: "manchmal", 3: "immer"},
    "sv": {1: "ibland", 3: "alltid"},
}

ISOLATED_WORD = {
    "it": "Senza legami", "en": "Unconnected", "es": "Sin vinculos",
    "fr": "Sans liens", "de": "Ohne Verbindung", "sv": "Utan kopplingar",
}

# Il pezzo accentato: uno solo, e la resa a parole lo dice, altrimenti chi
# ascolta perde l'unica enfasi che il tavolo sa portare.
ACCENT_WORD = {
    "it": "Il punto", "en": "The point", "es": "El punto",
    "fr": "Le point", "de": "Der Punkt", "sv": "Poangen",
}


class TavoloError(ValueError):
    """Un tavolo che non sta in piedi: contratto violato, non errore di sistema."""


def family_of(rel: str) -> str:
    """La famiglia non la dichiara nessuno: e' una proprieta' del verbo."""
    try:
        return REL_FAMILY[rel]
    except KeyError:
        raise TavoloError(f"connessione sconosciuta: {rel}") from None


class TavoloNode(BaseModel):
    model_config = ConfigDict(extra="ignore")

    id: str = Field(min_length=1, max_length=40)
    label: str = Field(min_length=1, max_length=MAX_LABEL)
    form: str = DEFAULT_FORM
    icon: str | None = Field(default=None, max_length=24)
    # L'enfasi della persona: il pezzo che conta. Uno solo per tavolo, come
    # nei diagrammi: due accenti non accentano niente.
    accent: bool = False
    # Chi lo ha messo li', e a che punto e'. Una proposta del modello non e'
    # contenuto del tavolo finche' la persona non l'ha guardata.
    by: Literal["person", "model"] = "person"
    state: Literal["live", "pending", "dropped"] = "live"
    # Dove sta sul tavolo. La disposizione e' della persona: il seme la calcola
    # una volta, poi nessuno la tocca piu'.
    x: float = 0.0
    y: float = 0.0

    @field_validator("form", mode="before")
    @classmethod
    def _known_form(cls, value):
        return value if value in NODE_FORMS else DEFAULT_FORM


class TavoloEdge(BaseModel):
    model_config = ConfigDict(extra="ignore", populate_by_name=True)

    source: str = Field(min_length=1, alias="from")
    target: str = Field(min_length=1, alias="to")
    rel: str
    label: str | None = Field(default=None, max_length=40)
    # I due modificatori. Non moltiplicano il vocabolario: lo qualificano.
    strength: int = Field(default=2, ge=1, le=3)
    hypothesis: bool = False
    by: Literal["person", "model"] = "person"
    state: Literal["live", "pending", "dropped"] = "live"

    @field_validator("rel")
    @classmethod
    def _known_rel(cls, value):
        family_of(value)
        return value

    @property
    def family(self) -> str:
        return REL_FAMILY[self.rel]

    @property
    def key(self) -> str:
        return f"{self.source}->{self.target}"


class TavoloGraph(BaseModel):
    model_config = ConfigDict(extra="ignore")

    title: str = Field(default="", max_length=MAX_TITLE)
    nodes: list[TavoloNode] = Field(default_factory=list, max_length=MAX_NODES)
    edges: list[TavoloEdge] = Field(default_factory=list, max_length=MAX_EDGES)


class TavoloProposal(BaseModel):
    """Cio' che il modello puo' dire. Solo aggiungere: togliere non gli compete."""

    model_config = ConfigDict(extra="ignore")

    add_nodes: list[TavoloNode] = Field(default_factory=list, max_length=MAX_PROPOSED)
    add_edges: list[TavoloEdge] = Field(default_factory=list, max_length=MAX_PROPOSED)
    note: str | None = Field(default=None, max_length=200)


def _validated(graph: TavoloGraph) -> TavoloGraph:
    known = {node.id for node in graph.nodes}
    if len(known) != len(graph.nodes):
        raise TavoloError("due nodi con lo stesso id")
    for edge in graph.edges:
        missing = {edge.source, edge.target} - known
        if missing:
            raise TavoloError(f"legame verso un nodo che non c'e': {', '.join(sorted(missing))}")
    if sum(1 for node in graph.nodes if node.accent) > 1:
        raise TavoloError("al massimo un pezzo accentato")
    return graph


def parse_graph(data: dict | str | TavoloGraph) -> TavoloGraph:
    """Valida un grafo (dict o JSON) e solleva `TavoloError` se non regge."""
    if isinstance(data, TavoloGraph):
        return _validated(data)
    try:
        if isinstance(data, str):
            return _validated(TavoloGraph.model_validate_json(data))
        return _validated(TavoloGraph.model_validate(data))
    except ValidationError as exc:
        raise TavoloError(str(exc)) from exc


def parse_proposal(data: dict | str | TavoloProposal) -> TavoloProposal:
    """Cio' che arriva dal modello, dict o JSON grezzo. Mai fidato: sempre validato."""
    if isinstance(data, TavoloProposal):
        return data
    try:
        if isinstance(data, str):
            return TavoloProposal.model_validate_json(data)
        return TavoloProposal.model_validate(data)
    except ValidationError as exc:
        raise TavoloError(str(exc)) from exc


def propose(graph: TavoloGraph, proposal: TavoloProposal) -> TavoloGraph:
    """Le mosse del modello entrano in sospeso e non toccano cio' che c'e'.

    Un id gia' presente non viene riscritto: il modello propone, e proporre di
    rinominare un nodo della persona senza chiederlo sarebbe scriverlo.
    """
    known = {node.id for node in graph.nodes}
    known_edges = {edge.key for edge in graph.edges}
    nodes = list(graph.nodes)
    edges = list(graph.edges)
    for node in proposal.add_nodes:
        if node.id in known:
            continue
        # L'accento e' della persona: il modello propone pezzi, non enfasi.
        nodes.append(node.model_copy(update={"by": "model", "state": "pending", "accent": False}))
        known.add(node.id)
    for edge in proposal.add_edges:
        if edge.key in known_edges:
            continue
        edges.append(edge.model_copy(update={"by": "model", "state": "pending"}))
        known_edges.add(edge.key)
    return _validated(TavoloGraph(title=graph.title, nodes=nodes, edges=edges))


def _settle(graph: TavoloGraph, ids: list[str], state: str) -> TavoloGraph:
    wanted = set(ids)
    if state == "live":
        alive = {node.id for node in graph.nodes if node.state == "live" or node.id in wanted}
        for edge in graph.edges:
            if edge.key not in wanted:
                continue
            orphan = {edge.source, edge.target} - alive
            if orphan:
                raise TavoloError(
                    f"il legame {edge.key} non regge senza {', '.join(sorted(orphan))}")
    nodes = [
        node.model_copy(update={"state": state}) if node.id in wanted and node.state == "pending"
        else node
        for node in graph.nodes
    ]
    edges = [
        edge.model_copy(update={"state": state}) if edge.key in wanted and edge.state == "pending"
        else edge
        for edge in graph.edges
    ]
    return TavoloGraph(title=graph.title, nodes=nodes, edges=edges)


def accept(graph: TavoloGraph, ids: list[str]) -> TavoloGraph:
    """Una proposta accettata diventa contenuto del tavolo, e resta del modello."""
    return _settle(graph, ids, "live")


def reject(graph: TavoloGraph, ids: list[str]) -> TavoloGraph:
    """Scartata, non cancellata: lo storico deve poter mostrare cosa fu offerto."""
    return _settle(graph, ids, "dropped")


def live(graph: TavoloGraph) -> TavoloGraph:
    """Il tavolo come contenuto: senza le proposte in sospeso e senza gli scarti."""
    nodes = [node for node in graph.nodes if node.state == "live"]
    alive = {node.id for node in nodes}
    edges = [
        edge for edge in graph.edges
        if edge.state == "live" and edge.source in alive and edge.target in alive
    ]
    return TavoloGraph(title=graph.title, nodes=nodes, edges=edges)


def from_idea_map(spec: dict) -> TavoloGraph:
    """La mappa di Idea diventa un tavolo senza passare da un modello.

    Ruoli e tipi di arco sono gia' un vocabolario: tradurli e' una tabella, e
    chiedere a un modello di rifare un lavoro gia' fatto ne perderebbe pezzi.
    Tutto quello che c'e' e' della persona: e' uscito dalla sua conversazione.
    """
    nodes = [
        TavoloNode(
            id=node["id"],
            label=node["label"],
            form=node.get("form") or FORM_FROM_ROLE.get(node.get("role") or "", DEFAULT_FORM),
            icon=node.get("icon"),
        )
        for node in spec.get("nodes", [])
    ]
    edges = [
        TavoloEdge.model_validate({
            "from": edge.get("from") or edge.get("source"),
            "to": edge.get("to") or edge.get("target"),
            "rel": REL_FROM_EDGE_KIND.get(edge.get("kind") or "link", "part-of"),
            "label": edge.get("label"),
        })
        for edge in spec.get("edges", [])
    ]
    return _validated(TavoloGraph(title=spec.get("title", ""), nodes=nodes, edges=edges))


def rendition(graph: TavoloGraph, lang: str = "it") -> str:
    """Il tavolo a parole: screen reader, TTS, ricerca nel PDF, Telegram.

    Passa da `live`: dire a voce una proposta che nessuno ha ancora accettato la
    farebbe passare per contenuto del tavolo.
    """
    code = (lang or "it").lower()[:2]
    words = REL_WORDS.get(code, REL_WORDS["en"])
    weights = STRENGTH_WORD.get(code, STRENGTH_WORD["en"])
    doubt = DOUBT_WORD.get(code, DOUBT_WORD["en"])
    content = live(graph)
    by_id = {node.id: node.label for node in content.nodes}

    relations = []
    for edge in content.edges:
        verb = edge.label.strip() if edge.label and edge.label.strip() else words[edge.rel]
        marks = [mark for mark in (weights.get(edge.strength), doubt if edge.hypothesis else None) if mark]
        tail = f" ({', '.join(marks)})" if marks else ""
        relations.append(f"{by_id[edge.source]} {verb} {by_id[edge.target]}{tail}")

    touched = {end for edge in content.edges for end in (edge.source, edge.target)}
    alone = [node.label for node in content.nodes if node.id not in touched]
    accented = next((node.label for node in content.nodes if node.accent), None)
    parts = ["; ".join(relations)] if relations else []
    if accented:
        parts.append(f"{ACCENT_WORD.get(code, ACCENT_WORD['en'])}: {accented}")
    if alone:
        parts.append(f"{ISOLATED_WORD.get(code, ISOLATED_WORD['en'])}: {', '.join(alone)}")
    body = ". ".join(part for part in parts if part)
    return f"{content.title}: {body}." if body else f"{content.title}."

File: frontend/src/lib/tavolo.ts
// Client del tavolo. Il vocabolario delle connessioni e' lo stesso di
// backend/tavolo.py: la famiglia si deriva dal verbo, non si dichiara, e forza
// e incertezza restano modificatori invece di moltiplicare i tipi.

// @ts-expect-error -- Node's direct TypeScript runner requires the extension.
import { apiFetch } from './auth.ts';

export type TavoloFamily = 'argument' | 'cause' | 'time' | 'part';
export type TavoloRel =
    | 'supports' | 'contradicts' | 'assumes' | 'needs-evidence'
    | 'causes' | 'hinders' | 'feeds-back'
    | 'then' | 'blocks' | 'if'
    | 'part-of' | 'example-of';
export type TavoloForm = 'concept' | 'action' | 'decision' | 'outcome';
// Il raggruppamento della persona. Non e' un canale condiviso come la famiglia
// dell'arco: vale quello che lei ci mette, e resta chiuso perche' una tavolozza
// libera su quaranta pezzi smette di raggruppare.
export type TavoloColor = 'green' | 'blue' | 'violet' | 'pink' | 'grey';
export const NODE_COLORS: TavoloColor[] = ['green', 'blue', 'violet', 'pink', 'grey'];
export type TavoloState = 'live' | 'pending' | 'dropped';
export type TavoloBy = 'person' | 'model';

export const REL_FAMILY: Record<TavoloRel, TavoloFamily> = {
    supports: 'argument', contradicts: 'argument', assumes: 'argument', 'needs-evidence': 'argument',
    causes: 'cause', hinders: 'cause', 'feeds-back': 'cause',
    then: 'time', blocks: 'time', if: 'time',
    'part-of': 'part', 'example-of': 'part',
};

export const FAMILIES: TavoloFamily[] = ['argument', 'cause', 'time', 'part'];

export const RELS_BY_FAMILY: Record<TavoloFamily, TavoloRel[]> = {
    argument: ['supports', 'contradicts', 'assumes', 'needs-evidence'],
    cause: ['causes', 'hinders', 'feeds-back'],
    time: ['then', 'blocks', 'if'],
    part: ['part-of', 'example-of'],
};

// Il colore e' il canale della famiglia: quattro tinte si distinguono a colpo
// d'occhio, tredici no. Petrolio e' il colore del marchio, esposto dalla scala
// indigo rimappata; ocra vuol dire movimento, come nel resto dell'app.
export const FAMILY_STROKE: Record<TavoloFamily, string> = {
    argument: 'var(--color-indigo-600)',
    cause: 'var(--color-slate-500)',
    time: 'var(--color-ochre-600)',
    part: 'var(--color-slate-300)',
};

export const INTENTS = ['what-is-missing', 'organize', 'connect', 'continue'] as const;
export type TavoloIntent = typeof INTENTS[number];

// I generi di schema. Gli id li conosce anche il server (backend/tavolo.py):
// il vocabolario di un genere arriva da la', queste sono le sue chiavi.
export const PRESET_IDS = ['workflow', 'causal', 'concept', 'argument', 'algorithm'] as const;
export type TavoloPresetId = typeof PRESET_IDS[number];

export interface TavoloPreset {
    id: TavoloPresetId;
    rels: TavoloRel[];
    forms: TavoloForm[];
    rankdir: 'LR' | 'TB';
    edge_label_required: boolean;
    prompts: string[];
    has_example: boolean;
}

export interface ComposeBody {
    preset: TavoloPresetId | null;
    prompt: string;
    counselor_id?: number;
    lang: string;
    base_index: number;
}

// Un prompt vuoto o un genere inventato non diventano una richiesta: il tetto
// del testo e la lista chiusa si controllano qui, dove costa niente.
export function composeBody(
    { preset, prompt, lang, index, counselorId }:
        { preset: TavoloPresetId | null; prompt: string; lang: string; index: number; counselorId?: number },
): ComposeBody | null {
    const text = prompt.trim().slice(0, 1200);
    if (!text) return null;
    if (preset !== null && !(PRESET_IDS as readonly string[]).includes(preset)) return null;
    return {
        preset,
        prompt: text,
        lang,
        base_index: index,
        ...(counselorId ? { counselor_id: counselorId } : {}),
    };
}

export interface TavoloNodeData {
    id: string;
    label: string;
    form: TavoloForm;
    icon?: string | null;
    color?: TavoloColor | null;
    // Il pezzo che conta: uno solo per tavolo, come nei diagrammi.
    accent?: boolean;
    by: TavoloBy;
    state: TavoloState;
    x: number;
    y: number;
}

export interface TavoloEdgeData {
    from: string;
    to: string;
    rel: TavoloRel;
    label?: string | null;
    strength: number;
    hypothesis: boolean;
    // Vale anche dall'altra parte: due punte invece di una.
    reciprocal?: boolean;
    by: TavoloBy;
    state: TavoloState;
}

export interface TavoloGraph {
    title: string;
    preset?: string | null;
    nodes: TavoloNodeData[];
    edges: TavoloEdgeData[];
}

export interface TavoloView {
    id: string;
    title: string | null;
    saved: boolean;
    origin_session_id: string | null;
    origin_instrument: string | null;
    index: number;
    graph: TavoloGraph;
    rendition: string | null;
    has_capture: boolean;
    note?: string | null;
}

export interface TavoloSummary {
    id: string;
    title: string | null;
    origin_instrument: string | null;
    has_capture: boolean;
    saved_at: string | null;
}

export const familyOf = (rel: TavoloRel): TavoloFamily => REL_FAMILY[rel];

export const edgeKey = (edge: TavoloEdgeData): string => `${edge.from}->${edge.to}`;

/** Cio' che conta come contenuto: le proposte in sospeso non sono ancora del tavolo. */
export function liveGraph(graph: TavoloGraph): TavoloGraph {
    const nodes = graph.nodes.filter((node) => node.state === 'live');
    const alive = new Set(nodes.map((node) => node.id));
    return {
        title: graph.title,
        nodes,
        edges: graph.edges.filter((edge) => edge.state === 'live' && alive.has(edge.from) && alive.has(edge.to)),
    };
}

export const pendingIds = (graph: TavoloGraph): string[] => [
    ...graph.nodes.filter((node) => node.state === 'pending').map((node) => node.id),
    ...graph.edges.filter((edge) => edge.state === 'pending').map(edgeKey),
];

async function json<T>(response: Response): Promise<T> {
    if (!response.ok) {
        const data = await response.json().catch(() => ({})) as { detail?: string };
        const error = new Error(data.detail || `HTTP ${response.status}`);
        (error as Error & { status?: number }).status = response.status;
        throw error;
    }
    return response.json() as Promise<T>;
}

const post = (path: string, body: unknown) => apiFetch(path, {
    method: 'POST', headers: { 'Content-Type': 'application/json' }, body: JSON.stringify(body),
});

export const createTavolo = (body: {
    session_id?: string; instrument?: string; title?: string; lang?: string;
    counselor_id?: number;
    // I due semi, esclusivi: la mappa di Idea arriva come contenuto, il testo
    // di una chat come proposta da accettare pezzo per pezzo.
    idea_map?: unknown; source_text?: string;
    preset?: TavoloPresetId | null;
}) => post('/api/tavolo', body).then((response) => json<TavoloView>(response));

export const fetchTavolo = (id: string) =>
    apiFetch(`/api/tavolo/${encodeURIComponent(id)}`).then((response) => json<TavoloView>(response));

export const listTavoli = () =>
    apiFetch('/api/tavolo').then((response) => json<TavoloSummary[]>(response));

// Se la funzione e' spenta gli endpoint rispondono 404, e i bottoni devono
// sparire invece di portare a un vicolo cieco. Non c'e' un canale per i flag
// verso il browser, quindi lo si chiede all'elenco: la promessa e' memorizzata
// qui perche' il bottone compare una volta per messaggio, e senza questa cache
// una conversazione lunga farebbe una richiesta per bolla.
let probe: Promise<boolean> | null = null;
export function tavoloEnabled(): Promise<boolean> {
    probe ??= apiFetch('/api/tavolo').then((response) => response.ok).catch(() => false);
    return probe;
}

export const writeTavolo = (id: string, graph: TavoloGraph, baseIndex: number) =>
    apiFetch(`/api/tavolo/${encodeURIComponent(id)}`, {
        method: 'PUT', headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({ graph, base_index: baseIndex }),
    }).then((response) => json<TavoloView>(response));

export const settleTavolo = (id: string, ids: string[], action: 'accept' | 'reject', baseIndex: number) =>
    post(`/api/tavolo/${encodeURIComponent(id)}/settle`, { ids, action, base_index: baseIndex })
        .then((response) => json<TavoloView>(response));

export const suggestTavolo = (id: string, body: {
    intent: TavoloIntent; counselor_id?: number; lang: string; base_index: number;
}) => post(`/api/tavolo/${encodeURIComponent(id)}/suggest`, body).then((response) => json<TavoloView>(response));

export const fetchPresets = (lang: string): Promise<TavoloPreset[]> =>
    apiFetch(`/api/tavolo/presets?lang=${encodeURIComponent(lang)}`)
        .then((response) => json<{ presets: TavoloPreset[] }>(response))
        .then((body) => body.presets);

export const fetchPresetExample = (id: TavoloPresetId, lang: string): Promise<TavoloGraph> =>
    apiFetch(`/api/tavolo/presets/${id}/example?lang=${encodeURIComponent(lang)}`)
        .then((response) => json<{ graph: TavoloGraph }>(response))
        .then((body) => body.graph);

export const composeTavolo = (id: string, body: ComposeBody): Promise<TavoloView & { note?: string | null }> =>
    post(`/api/tavolo/${encodeURIComponent(id)}/compose`, body)
        .then((response) => json<TavoloView & { note?: string | null }>(response));

export const saveTavolo = (id: string, title: string, lang: string) =>
    post(`/api/tavolo/${encodeURIComponent(id)}/save`, { title, lang })
        .then((response) => json<TavoloView>(response));

/** La cattura e' facoltativa: se fallisce, il tavolo resta salvato con la sua resa a parole. */
export async function uploadCapture(id: string, png: Blob): Promise<boolean> {
    const form = new FormData();
    form.append('file', png, `${id}.png`);
    const response = await apiFetch(`/api/tavolo/${encodeURIComponent(id)}/capture`, {
        method: 'POST', body: form,
    });
    return response.ok;
}
